File: src/main/java/uk/laxd/dndSimulator/event/EncounterEventFactory.java
package uk.laxd.dndSimulator.event;

import org.springframework.stereotype.Component;
import uk.laxd.dndSimulator.action.Action;
import uk.laxd.dndSimulator.action.MeleeAttackAction;
import uk.laxd.dndSimulator.action.Turn;

@Component
public class EncounterEventFactory {

    public EncounterEvent createTurnStartEvent(Turn turn) {
        EncounterEvent event = new EncounterEvent();
        event.type = EncounterEventType.TURN_START;
        event.actor = turn.getCharacter();

        return event;
    }

    public EncounterEvent createNewActionEvent(Action action) {
        // TODO: Handle movement, reactions, spells etc

        if(action.getEventType() == EncounterEventType.MELEE_ATTACK) {
            MeleeAttackAction meleeAttackAction = (MeleeAttackAction) action;

            EncounterEvent event = new EncounterEvent();
            event.actor = meleeAttackAction.getPerformer();
            event.target = meleeAttackAction.getTarget();
            event.amount = meleeAttackAction.getAttackDamage();
            event.weapon = meleeAttackAction.getWeapon();
            event.type = EncounterEventType.MELEE_ATTACK;
            event.eventOutcome = meleeAttackAction.getOutcome();

            return event;
        }
        else {
            EncounterEvent event = new EncounterEvent();
            event.type = action.getEventType();

            return event;
        }
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/event/EncounterEventFactory.kt
package uk.laxd.dndSimulator.event

import org.springframework.stereotype.Component
import uk.laxd.dndSimulator.action.Action
import uk.laxd.dndSimulator.action.Turn
import uk.laxd.dndSimulator.action.MeleeAttackAction
import uk.laxd.dndSimulator.character.Character

@Component
class EncounterEventFactory {

    fun createTurnStartEvent(turn: Turn): EncounterEvent {
        return TurnStartEvent(turn.character)
    }

    fun createCharacterDeathEvent(character: Character, killedBy: Character): EncounterEvent {
        return DeathEvent(character, killedBy)
    }

    fun createNewActionEvent(action: Action): EncounterEvent {
        // TODO: Handle movement, reactions, spells etc

        return when (action.eventType) {
            EncounterEventType.MELEE_ATTACK -> {
                val meleeAttackAction = action as MeleeAttackAction
                MeleeAttackEvent(
                    meleeAttackAction.actor,
                    meleeAttackAction.weapon,
                    meleeAttackAction.target,
                    meleeAttackAction.attackRollResult,
                    meleeAttackAction.damageRollResult,
                    meleeAttackAction.outcome,
                    meleeAttackAction.attackDamage
                )
            }
            else -> {
                GeneralEncounterEvent(action.actor, action.eventType)
            }
        }
    }
}